sage/group_prover.sage

```diff
@@ -177,6 +177,30 @@ class constraints:
   def __repr__(self):
     return "%s" % self
 
+def normalize_factor(p):
+  """Normalizes the sign of primitive polynomials (as returned by factor())
+
+  This function ensures that the polynomial has a positive leading coefficient.
+
+  This is necessary because recent sage versions (starting with v9.3 or v9.4,
+  we don't know) are inconsistent about the placement of the minus sign in
+  polynomial factorizations:
+  ```
+  sage: R.<ax,bx,ay,by,Az,Bz,Ai,Bi> = PolynomialRing(QQ,8,order='invlex')
+  sage: R((-2 * (bx - ax)) ^ 1).factor()
+  (-2) * (bx - ax)
+  sage: R((-2 * (bx - ax)) ^ 2).factor()
+  (4) * (-bx + ax)^2
+  sage: R((-2 * (bx - ax)) ^ 3).factor()
+  (8) * (-bx + ax)^3
+  ```
+  """
+  # Assert p is not 0 and that its non-zero coeffients are coprime.
+  # (We could just work with the primitive part p/p.content() but we want to be
+  # aware if factor() does not return a primitive part in future sage versions.)
+  assert p.content() == 1
+  # Ensure that the first non-zero coefficient is positive.
+  return p if p.lc() > 0 else -p
```

Comment:
What about using `p / p.lc()`? That will always guarantee a result with leading coefficient 1, even if there is more than 1/-1 variability.

Reply:
That seems possible but by asserting that the [content](https://doc.sagemath.org/html/en/reference/polynomial_rings/sage/rings/polynomial/multi_polynomial.html#sage.rings.polynomial.multi_polynomial.MPolynomial.content) is 1 we shouldn't have more than 1/-1 variability.

Reply:
Sure, but if you use `p / p.lc()`, there is no need for such an assertion.

Reply:
Ok. It seemed slightly more appealing to me to just patch the sage bug but \*shrug\*

Reply:
> What about using `p / p.lc()`? That will always guarantee a result with leading coefficient 1, even if there is more than 1/-1 variability.

That seems to work, though this changes the representation further. `p/p.content() if p.lc() > 0 else -p/p.content()` will be closer to what we have now (and return a primitive polynomial).

But yeah, see the comment. It was an intentional decision to add the assertion. I think if that assertion fails, it may be the case that sage did larger changes and we may want to investigate this then.

But I can change it if you prefer.

Reply:
No problem.